### ISSUE TITLE
suspicious/malformed csv records produced by MycoPortal Symbiota portal

### ISSUE THREAD

**Issue by jhpoelen** [BioKIC/Symbiota#758]
Hi!

First, thanks for work to keep Symbiota up and running!

While working on some GloBI improvements, I so happen to use digital catalogues associated with collections in Mycoportal as a sample corpus.

This is why I noticed some suspicious entries in the GloBI review of possible interaction claims (e.g., fungal hosts) at https://depot.globalbioticinteractions.org/reviews/globalbioticinteractions/mycoportal/ .

It appears that some of the occurrence data inside Darwin Core archives is at odds with CSV formatting.  The GBIF DwC-A parser is choking on it. This likely prevents this and/or records from being re-used. 

For example,  today's (2023-12-07) copy of University of Michigan Herbarium retrieved via

```
<https://mycoportal.org/portal/content/dwca/MICH_DwC-A.zip> <http://purl.org/pav/hasVersion> <hash://sha256/199fbf1adc3b8f8d1e19e2c97e151b2316787c24876bbbb0aa1a1f2957e86061> <urn:uuid:e8098640-a508-44cc-9e70-9c75fb3def61> .
```

contains a suspected malformed csv line in the occurrences.csv table on line L114943 .

You can find the offending line via

[<https://linker.bio/line:zip:hash://sha256/199fbf1adc3b8f8d1e19e2c97e151b2316787c24876bbbb0aa1a1f2957e86061!/occurrences.csv!/L1,L114943>](https://linker.bio/line:zip:hash://sha256/199fbf1adc3b8f8d1e19e2c97e151b2316787c24876bbbb0aa1a1f2957e86061!/occurrences.csv!/L1,L114943)

or below

```csv
id,institutionCode,collectionCode,ownerInstitutionCode,collectionID,basisOfRecord,occurrenceID,catalogNumber,otherCatalogNumbers,higherClassification,kingdom,phylum,class,order,family,scientificName,taxonID,scientificNameAuthorship,genus,subgenus,specificEpithet,verbatimTaxonRank,infraspecificEpithet,taxonRank,identifiedBy,dateIdentified,identificationReferences,identificationRemarks,taxonRemarks,identificationQualifier,typeStatus,recordedBy,recordNumber,eventDate,year,month,day,startDayOfYear,endDayOfYear,verbatimEventDate,occurrenceRemarks,habitat,fieldNumber,informationWithheld,dataGeneralizations,dynamicProperties,associatedOccurrences,associatedTaxa,reproductiveCondition,establishmentMeans,lifeStage,sex,individualCount,preparations,country,stateProvince,county,municipality,locality,locationRemarks,decimalLatitude,decimalLongitude,geodeticDatum,coordinateUncertaintyInMeters,verbatimCoordinates,georeferencedBy,georeferenceProtocol,georeferenceSources,georeferenceVerificationStatus,georeferenceRemarks,minimumElevationInMeters,maximumElevationInMeters,minimumDepthInMeters,maximumDepthInMeters,verbatimDepth,verbatimElevation,disposition,language,recordEnteredBy,modified,rights,rightsHolder,accessRights,recordId,references
2078049,MICH,Fungi,,0cd2551b-8166-4c05-a0c9-5c8712ce0eb8,PreservedSpecimen,425c0ac4-df41-4000-ab0e-e91189ee0d16,178201,,Fungi|Basidiomycota|Agaricomycotina|Agaricomycetes|Agaricomycetidae|Agaricales|Mycenaceae,Fungi,Basidiomycota,Agaricomycetes,Agaricales,Mycenaceae,Mycena,16268,"(Pers.) Roussel",Mycena,,,,,Genus,,,,,,,,"W. B. Cooke; V. G. Cooke",40093,1968-09-21,1968,9,21,265,,,"The collection is stored in the indeterminate Mycena box with the preliminary determination ""Mycena \\".",,,,,,,,,,,,,,USA,Ohio,Preble,,"Hueston Woods.",,,,,,,,,,,,,,,,,,,,obreliza,"2013-10-24 00:00:00",http://creativecommons.org/licenses/by-nc/3.0/,,,urn:uuid:425c0ac4-df41-4000-ab0e-e91189ee0d16,https://mycoportal.org/portal/collections/individual/index.php?occid=2078049
```

or the attached csv snippet 
[MICH_DwC_L114943.csv](https://github.com/BioKIC/Symbiota/files/13605755/MICH_DwC_L114943.csv)
 as retrieved from above.

the associated record references are:

https://mycoportal.org/portal/collections/individual/index.php?occid=2078049
occurrenceId: 425c0ac4-df41-4000-ab0e-e91189ee0d16
catalogue number:  178201

The root cause is the escaping of the double quotes inside the notes field with problematic section highlighted -

> "The collection is stored in the indeterminate Mycena box with the preliminary determination **""Mycena \\"**."

A proposed fix for the example above would be:

> "The collection is stored in the indeterminate Mycena box with the preliminary determination **""Mycena ""**."

Note how the ```""``` is used consistently to escape ```"``` . See also attached version of suggested fix. 
[MICH_DwC_L114943-suggested-fix.csv](https://github.com/BioKIC/Symbiota/files/13605772/MICH_DwC_L114943-suggested-fix.csv)

Please confirm that Symbiota is expected to export only valid csv files inside their DwC-A.

Thank you, and please let me know if there's something I've missed. Probably not the first time this stuff came up.

PS For completeness, I've included the exact version of the DwC-A under review as retrieved from the MycoPortal on 2023-12-07 at https://linker.bio/hash://sha256/199fbf1adc3b8f8d1e19e2c97e151b2316787c24876bbbb0aa1a1f2957e86061.zip .

**Comment by themerekat** [comment]
Thanks for your feedback @jhpoelen ! It seems like what you're looking for would be some sort of CSV or Darwin Core Archive validator, correct?
I've run into a similar problem with malformed DwC-As preventing publication of the data to GBIF. Once the special characters (or whatever they are) are removed, everything runs smoothly. I imagine these aberrations arise from manual data uploads somewhere in the past, perhaps as tsvs.

**Comment by jhpoelen** [comment]
@themerekat thanks for your prompt reply. 

I'd be curious to see how this particular malformed csv was generated. . . (a) was the DwC-A generated by Symbiota? (b) Or was is manually put together and uploaded as is? 

If (a), then escaping values written into csv fields should avoid these issues.

if (b), then a monitoring solution might help to detect these kinds of issues and notify the curators. 

So, would you happen to know what process generated the line referred to in:

https://linker.bio/line:zip:hash://sha256/199fbf1adc3b8f8d1e19e2c97e151b2316787c24876bbbb0aa1a1f2957e86061!/occurrences.csv!/L1,L114943

?

**Comment by themerekat** [comment]
No, I don't. The record was created in 2013 and could have been from some sort of upload (tsv file, IPT, Specify import, added to backend manually, etc.), but I cannot tell.

**Comment by jhpoelen** [comment]
@themerekat just to be sure . . . I am referring to the specific line written into the occurrences.csv table of a specific DwC-A, not when the record was first created. Aren't the DwC-A zip create and packaged by Symbiota through some kind of export process?

**Comment by themerekat** [comment]
Oh, you mean what process massaged that into the Darwin Core Archive? I think that would be our DarwinCore Archiver processes (base here: https://github.com/BioKIC/Symbiota/blob/master/classes/DwcArchiverCore.php, but there are many ancillary classes)

**Comment by jhpoelen** [comment]
timeline of https://mycoportal.org/portal/collections/individual/index.php?occid=2078049

2013-??-?? - record first created
[...] many curation updates [...]
2023-??-?? - record exported to DwC-A  with signature https://linker.bio/hash://sha256/199fbf1adc3b8f8d1e19e2c97e151b2316787c24876bbbb0aa1a1f2957e86061.zip
2023-12-07 - Record updated by curator
2023-12-08 - record re-exported to DwC-A with signature X 

So, my question is - what is the program that generates the DwC-A zip files?

**Comment by themerekat** [comment]
> > Oh, you mean what process massaged that into the Darwin Core Archive? I think that would be our DarwinCore Archiver processes (base here: https://github.com/BioKIC/Symbiota/blob/master/classes/DwcArchiverCore.php, but there are many ancillary classes)
> 
> Yes! So, somewhere in there should be functionality that takes a field from the database and writes it into the csv files that are included in the DwC-A. That functionality should ensure to escape csv special characters, like the double quote `"`. For some reason, this does not appear to happen. Can you confirm that no csv escaping is currently implemented in the DwCArchiverCore.php and associated classes?

I'm sure there are some sort of checks/escapes, but we'd have to look into it more closely to determine if this "\\" is an outlier / non-escaped case.

**Comment by jhpoelen** [comment]
Thanks to manual curation and expertise of @themerekat , Alison, Chelsea,  and Phil, the immediate issues with the MycoPortal collections have been resolved, making thousands of records available for indexing through GBIF, GloBI and other indexing infrastructures. 

However, I hope that Symbiota's DwC-A writer will be improved such that valid CSV are generated, regardless of what text may sit in record fields.
